authz/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+#![warn(
+    missing_copy_implementations,
+    missing_docs,
+    clippy::explicit_iter_loop,
+    clippy::future_not_send,
+    clippy::use_self,
+    clippy::clone_on_ref_ptr,
+    clippy::todo,
+    clippy::dbg_macro
+)]
+#![allow(rustdoc::private_intra_doc_links)]
+
+use async_trait::async_trait;
+use generated_types::influxdata::iox::authz::v1 as proto;
+use hyper::Request;
+use snafu::Snafu;
+
+mod permission;
+pub use permission::{Action, Permission, Resource};
+
+/// An authorizer is used to validate the associated with
+/// an authorization token that has been extracted from a request.
+#[async_trait]
+pub trait Authorizer: std::fmt::Debug + Send + Sync {
+    /// Determine the permissions allowed by a request token.
+    ///
+    /// The returned list of permissions is the intersection of the permissions
+    /// requested and the permissions associated with the token. An error
+    /// will only be returned if there is a failure processing the token.
+    /// An invalid token is taken to have no permissions, so these along
+    /// with tokens that match none of the requested permissions will return
+    /// empty permission sets.
+    async fn allow(
+        &self,
+        token: &[u8],
+        perms: &[Permission],
+    ) -> Result<Vec<Permission>, VerifyError>;
+}
+
+/// Authorizer implementation using influxdata.iox.authz.v1 protocol.
+#[derive(Clone, Debug)]
+pub struct IoxAuthorizer {
+    client:
+        proto::iox_authorizer_service_client::IoxAuthorizerServiceClient<tonic::transport::Channel>,
+}
+
+impl IoxAuthorizer {
+    /// Attempt to create a new client by connecting to a given endpoint.
+    pub fn connect_lazy<D>(dst: D) -> Result<Self, Box<dyn std::error::Error>>
+    where
+        D: TryInto<tonic::transport::Endpoint> + Send,
+        D::Error: Into<tonic::codegen::StdError>,
+    {
+        let ep = tonic::transport::Endpoint::new(dst)?;
+        let client = proto::iox_authorizer_service_client::IoxAuthorizerServiceClient::new(
+            ep.connect_lazy(),
+        );
+        Ok(Self { client })
+    }
+}
+
+#[async_trait]
+impl Authorizer for IoxAuthorizer {
+    async fn allow(
+        &self,
+        token: &[u8],
+        perms: &[Permission],
+    ) -> Result<Vec<Permission>, VerifyError> {
+        let req = proto::AuthorizeRequest {
+            token: token.to_vec(),
+            permissions: perms.iter().filter_map(|p| p.try_into().ok()).collect(),
+        };
+        let mut client = self.client.clone();
+        let resp = client.authorize(req).await?;
+        Ok(resp
+            .into_inner()
+            .permissions
+            .iter()
+            .filter_map(|p| p.try_into().ok())
+            .collect())
+    }
+}
+
+/// Communication error when verifying a token.
+#[derive(Debug, Snafu)]
+#[snafu(display("token verification not possible: {msg}"))]
+pub struct VerifyError {
+    msg: String,
+    source: Box<dyn std::error::Error + Send + Sync>,
```

Comment:
I think you'll probably want:

```suggestion
    source: Box<dyn std::error::Error + Send + Sync + 'static>,
```

Unless your planning on returning refs. It makes working with async a lot easier when the types are 'static :+1: